homeassistant/components/xiaomi_miio/number.py

```diff
@@ -0,0 +1,214 @@
+                        f"{number.short_name}_{config_entry.unique_id}",
+                        number,
+                        coordinator,
+                    )
+                )
+
+    async_add_entities(entities)
+
+
+class XiaomiNumber(XiaomiCoordinatedMiioEntity, NumberEntity):
+    """Representation of a generic Xiaomi attribute selector."""
+
+    def __init__(self, name, device, entry, unique_id, number, coordinator):
+        """Initialize the generic Xiaomi attribute selector."""
+        super().__init__(name, device, entry, unique_id, coordinator)
+        self._state = None
+        self._attr_icon = number.icon
+        self._attr_unit_of_measurement = number.unit_of_measurement
+        self._attr_min_value = number.min
+        self._attr_max_value = number.max
+        self._attr_step = number.step
+        self._supported_features = 0
+        self._device_features = 0
+        self._state_attrs = {}
+        self._controller = number
+        self._enum_class = None
+        self._attributes = None
+        self._value = None
+
+    @property
+    def value(self):
+        """Return the current numbered value."""
+        return self._value
+
+    @property
+    def available(self):
+        """Return the number controller availability."""
+        if (
+            super().available
+            and not self.coordinator.data.is_on
+            and not self._controller.available_with_device_off
+        ):
+            return False
+        return super().available
+
+    @staticmethod
+    def _extract_value_from_attribute(state, attribute):
+        value = getattr(state, attribute)
+        if isinstance(value, Enum):
+            return value.value
+
+        return value
+
+    async def async_set_value(self, value):
+        """Set an option of the miio device."""
+        if not self.available:
+            return
+        if (
+            self.min_value
+            and value < self.min_value
+            or self.max_value
+            and value > self.max_value
+        ):
+            _LOGGER.warning(
+                "Value %s not a valid %s within the range %s - %s",
+                value,
+                self.name,
+                self.min_value,
+                self.max_value,
+            )
+            return
+        method = getattr(self, SERVICE_TO_METHOD[self._controller.service]["method"])
+        if await method(value):
+            self._value = value
+            self.async_write_ha_state()
+
+
+class XiaomiAirHumidifierNumber(XiaomiNumber):
+    """Representation of a Xiaomi Air Humidifier selector."""
+
+    def __init__(self, name, device, entry, unique_id, controller, coordinator):
+        """Initialize the plug switch."""
+        super().__init__(name, device, entry, unique_id, controller, coordinator)
+        if self._model in [MODEL_AIRHUMIDIFIER_CA1, MODEL_AIRHUMIDIFIER_CB1]:
+            self._device_features = FEATURE_FLAGS_AIRHUMIDIFIER_CA_AND_CB
```

Comment:
We already check this in the platform setup above. Can we just pass the device features to the entity?

Reply:
Cleaned up